x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java

```diff
@@ -274,11 +296,147 @@ public SqlFunctionRegistry() {
                 def(StY.class, StY::new, "ST_Y"),
                 def(StZ.class, StZ::new, "ST_Z")
             },
-        // Special
-            new FunctionDefinition[] {
+            // Special
+            new FunctionDefinition[]{
                 def(Score.class, Score::new, "SCORE")
             }
         };
     }
 
+    /**
+     * Builder for creating SQL-specific functions.
+     * All other definitions defined here end up being translated to this form.
+     */
+    protected interface SqlFunctionBuilder {
+        Function build(Source source, List<Expression> children, Configuration cfg, Boolean distinct);
+    }
+
+    /**
+     * Main method to register a function.
+     */
+    @SuppressWarnings("overloads")
+    protected static FunctionDefinition def(Class<? extends Function> function,
+                                            SqlFunctionBuilder builder,
+                                            boolean datetime,
+                                            String... names) {
+        Check.isTrue(names.length > 0, "At least one name must be provided for the function");
+        String primaryName = names[0];
+        List<String> aliases = Arrays.asList(names).subList(1, names.length);
+        FunctionDefinition.Builder realBuilder = (uf, cfg, extras) -> {
+            try {
+                return builder.build(uf.source(), uf.children(), cfg, asBool(extras));
+            } catch (QlIllegalArgumentException e) {
+                throw new ParsingException(uf.source(), "error building [" + primaryName + "]: " + e.getMessage(), e);
+            }
+        };
+        return new SqlFunctionDefinition(primaryName, unmodifiableList(aliases), function, datetime, realBuilder);
+    }
+
+    private static Boolean asBool(Object[] extras) {
+        if (CollectionUtils.isEmpty(extras)) {
+            return null;
+        }
+        if (extras.length != 1 || (extras[0] instanceof Boolean) == false) {
+            throw new SqlIllegalArgumentException("Invalid number and types of arguments given to function definition");
```

Comment:
Does it make sense to be more explicit in the error message here? The message says it's an invalid number and type of arguments but it doesn't say what is expecting.